contrib/envconfig/client_config.go

```diff
@@ -0,0 +1,210 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package envconfig contains utilities to load configuration from files and/or environment variables.
+//
+// WARNING: Environment configuration is currently experimental.
+package envconfig
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/converter"
+)
+
+// ClientConfig represents a client config file.
+//
+// WARNING: Environment configuration is currently experimental.
+type ClientConfig struct {
+	Profiles map[string]*ClientConfigProfile
+}
+
+// ClientConfigProfile is profile-level configuration for a client.
+//
+// WARNING: Environment configuration is currently experimental.
+type ClientConfigProfile struct {
+	Address   string
+	Namespace string
+	APIKey    string
+	TLS       *ClientConfigTLS
+	Codec     *ClientConfigCodec
+	// When loading from TOML and env var, or writing to TOML, the keys are lowercased and hyphens are replaced with
+	// underscores. This is used for deduplicating/overriding too, so manually set values that are not normalized may
+	// not get overridden with [ClientConfigProfile.ApplyEnvVars].
+	GRPCMeta map[string]string
+}
+
+// ClientConfigTLS is TLS configuration for a client.
+//
+// WARNING: Environment configuration is currently experimental.
+type ClientConfigTLS struct {
+	Disabled                bool
```

Comment:
So the default will be TLS enabled?

Reply:
Or are we treating `nil` different then the zero value?

Reply:
> So the default will be TLS enabled?

No, the default is "depends on other fields". If you give an API key the default is enabled, but if you don't the default is disabled (but you can make enabled with any TLS value including TLS itself being set). So this is effectively for those rare cases that have an API key but want to explicitly disable TLS (e.g. self-hosted using JWTs). This is what allows us to just accept an API key env var for cloud and assume TLS.

Reply:
Unset/false here is "not disabled", not necessarily "enabled". I agree it can be confusing, but I think any alternative is more confusing.